Add support for fetching rules with API client

bobmshannon opened the pull request:
This PR adds support for fetching [rules](https://prometheus.io/docs/prometheus/latest/querying/api/#rules) via the `/api/v1/rules` endpoint using the API client. Currently, the API client exposes no way to do this and it would be nice to have for external systems that wish to retrieve this information.

cc @beorn7

bobmshannon commented on api/prometheus/v1/api.go:
```diff
@@ -139,6 +165,38 @@ type SnapshotResult struct {
+type RuleGroup struct {
+	Name     string  `json:"name"`
+	File     string  `json:"file"`
+	Rules    []Rule  `json:"rules"`
```
Since a Rule can be either `recording` or `alerting`, I'm not sure if we actually want to use a stronger type here with two fields `RecordingRules` and `AlertingRules`, or put everything in the same list to preserve the rule order like https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L721 does.

krasi-georgiev replied:
I will let @simonpasquier 	and @mxinden comment on that.

beorn7 commented:
That one is for @krasi-georgiev . :smile:

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -67,6 +82,15 @@ const (
+	RuleTypeAlerting  RuleType = "alerting"
+
+	// Possible values for RuleHealth.
+	RuleHealthGood    = "ok"
```
Why not import them from here? https://github.com/prometheus/prometheus/blob/master/rules/manager.go#L46

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -67,6 +82,15 @@ const (
 	HealthBad     HealthStatus = "down"
+
+	// Possible values for RuleType.
+	RuleTypeRecording RuleType = "recording"
```
Mhh, its a bummer that I didn't expose these as constants when I added this. Sorry about that. Feel free to propose a patch on prometheus/prometheus.

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -139,6 +165,38 @@ type SnapshotResult struct {
+	Groups []RuleGroup `json:"groups"`
+}
+
+type RuleGroup struct {
```
Why not use https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L718 ?

bobmshannon replied:
The primary reasons for not not importing `RuleGroup` from the `v1` API package referenced above:
-`RuleGroup` models the list of rules as `interface{}` - this requires the client to use type assertion on a private type to determine whether a given rule is a recording or alerting rule.
- It adds an external dependency that the client will need to import in order to use this API client package but I think we can encapsulate the types to mitigate this concern.

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -139,6 +165,38 @@ type SnapshotResult struct {
+	Interval float64 `json:"interval"`
+}
+
+type Rule struct {
```
The name `Rule` is misleading, as recording rules do not expose e.g. `Alerts`.

mxinden replied:
Again, could we import structs from prometheus/prometheus instead?

bobmshannon replied:
I think we should just create two new types here: `RecordingRule` and `AlertingRule`. Is there any reason why order needs to be preserved by putting both types of rules in the same array?

If we create these two new types, I think we'll need some changes to the imported structs, for example `recordingRule` and `alertingRule` are not exported:
https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L730
https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L743

bobmshannon commented:
I'm happy make the changes to this PR such that the types are imported from `prometheus/prometheus`, however I'll also need to make a patch to export some types so that they can be consumed here. Additionally, I think I'd prefer to keep the types encapsulated as much as possible so that clients don't need to import a whole bunch of external packages just to use the API client. Let me know what you all think.

As for using OpenAPI to generate the client code, agreed it would be a very nice place to get to.

bobmshannon commented:
@mxinden Do you mind having another look at this one? I addressed some of the comments and added new fields `AlertRules` and `RecordingRules` to the `RuleGroup` struct to provide stronger typing and reduce ambiguity from a client perspective.

I think that importing the types at this point may be more trouble than it's worth, and we should probably just plan to supercede this client with the OpenAPI generated one, however I could be wrong about this.

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -169,6 +240,63 @@ type queryResult struct {
+	var recordingRules []RecordingRule
+
+	for _, rule := range v.Rules {
+		// Because both recording and alerting rules are stored in the same
```
They are stored in the same JSON array as the ordering among them matters (See https://github.com/prometheus/prometheus/pull/4318#discussion_r202675025 for more details). That way users can understand which rules depend on which.

I am not sure this is a blocker. But probably worth mentioning in a comment somewhere. What do you think?

bobmshannon replied:
Thanks for pointing me to that.

To keep things consistent, I've modified the client to preserve rule order as returned by the API, and added some documentation on this behavior regarding how to use a type switch to access each recording or alerting rule.

mxinden commented:
On the one hand given the fact that this client is already tightly coupled with the `prometheus/prometheus` structs (e.g. when we change something in `prometheus/prometheus` this will likely break this client), I think this dependency should also be expressed in the code, hence importing it.

On the other hand according to the README of this repository we don't give any stability guarantees in regards to the client. Thereby I am fine with going either way.

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -139,6 +166,63 @@ type SnapshotResult struct {
+//   default:
+//   	fmt.Printf("unknown rule type %s", v)
+//   }
+type Rules []interface{}
```
Where is this type used? Am I missing something?

bobmshannon replied:
Nice catch, fixed.

mxinden commented on api/prometheus/v1/api.go:
```diff
@@ -169,6 +253,56 @@ type queryResult struct {
+		if !ok {
+			return errors.New("rule has no type field")
+		}
+		ruleJSON, err := json.Marshal(rule)
```
I am a bit confused here. You are marshaling a rule to unmarshal it?

How about unmarshaling the rule into something like this first:

```golang
struct {
  type string
}
```

and depending on `type` unmarshal as a recording or an alerting rule?

bobmshannon replied:
With that approach I think we'd still need to get the complete byte or JSON encoding of each individual rule (by using MarshalJSON) in order to unmarshal them into a typed recording or alerting rule. That said, I think most of the complexity here is from decoding the mixed type slice and there may not be a super elegant way to do it in Go.

bobmshannon replied:
I've made an update to this PR that should be easier to read and is more in-line with the approach suggested in this comment.

bobmshannon commented:
@mxinden I've made some adjustments to this PR based on the most recent comments, please let me know what you think.

bobmshannon commented:
@beorn7 @mxinden Just wanted to circle back on this one, is there anything I can do to help push this one through? It would be really nice to have this functionality in the API client for my use-case and for any external system that fetches alert and rules data from Prometheus. Thanks in advanced. :)

beorn7 commented:
I'd like to see @mxinden and @krasi-georgiev to have the final look here.

krasi-georgiev commented:
yeah waiting for Max for the final input.

bobmshannon commented:
Great, thanks!

krasi-georgiev commented:
Thanks
